### PR TITLE
Add IE proxy setting support to winhttp stagers

### PR DESCRIPTION
**Pull request by OJ**
## Description

This PR adds IE proxy configuration support to the `reverse_winhttps` for x64 and x86. This code is enabled by default via the `PayloadProxyIE` setting in the datastore (default is `true`). This makes the size of the payload bigger by quite a few characters thanks to the requirements of the APIs in question, and so can be disabled by setting this value to `false`.

Note that Meterpreter binaries should be updated as well, as they contain a small fix that should be in place. That can be found here: https://github.com/rapid7/meterpreter/pull/166

This PR also relies on the UUID payload work (as it piggypacks off the new stagers). So #5367 needs to be landed first.

This means that the stagers should be able to support IE proxy _and_ SSL certificate validation _at the same time_.

hooray
## Verification
- [x] Land #5367
- [x] Land https://github.com/rapid7/meterpreter/pull/166
- [x] x86 `reverse_winhttp/s` works with IE proxy auto configuration url (PAC file) with and without SSL verification
- [x] x86 `reverse_winhttp/s` works with IE proxy set directly with and without SSL verification
- [x] x64 `reverse_winhttp/s` works with IE proxy auto configuration url (PAC file) with and without SSL verification
- [x] x64 `reverse_winhttp/s` works with IE proxy set directly with and without SSL verification
- [x] All payloads function as they did before when the `PayloadProxyIE` datastore option is set to `false`
- [x] Update the payload cache sizes before landing

Each of these proxy settings are modifiable under the context of the current user via the IE Internet Options dialog.

### CODE REVIEW

**Comment by bcook-r7** [comment]
Hi @OJ can you rebase or merge master so the diff here is manageable?

**Comment by OJ** [comment]
There we go, that should be a little more digestable! Thanks.

**Comment by kernelsmith** [comment]
best gif ever

**Comment by OJ** [comment]
Hey @bcook-r7, I know you're AFK at the mo, but when you get back please let me know if you have any more holdups that you need me to address for this. Thanks!

**Comment by OJ** [comment]
Bump! :)

File: lib/msf/core/payload/windows/reverse_winhttp.rb [hunk 1]
```diff
@@ -117,9 +123,22 @@ def asm_reverse_winhttp(opts={})
     retry_count       = [opts[:retry_count].to_i, 1].max
     verify_ssl        = nil
     encoded_cert_hash = nil
-    encoded_url       = asm_generate_wchar_array(opts[:url])
+    encoded_uri       = asm_generate_wchar_array(opts[:uri])
     encoded_host      = asm_generate_wchar_array(opts[:host])
```
Comment (bcook-r7) [inline lib/msf/core/payload/windows/reverse_winhttp.rb]: is there any good way for winhttp/https to share this code? it seems like a lot to duplicate between them.

Comment (OJ) [inline lib/msf/core/payload/windows/reverse_winhttp.rb]: Yes this should happen. I'll do something about this today.

Comment (OJ) [inline lib/msf/core/payload/windows/reverse_winhttp.rb]: Been pondering this a bit. There's a lot of config-based duplication all over this stuff, including the existing stagers. I'm thinking that I should instead create another PR after this that tidies up a bunch of stuff, and not just for these two stagers.

Are you cool with that @bcook-r7 ?

Comment (bcook-r7) [inline lib/msf/core/payload/windows/reverse_winhttp.rb]: Oh, absolutely.

**Comment by bcook-r7** [comment]
These all look OK :) I kept running into mistakes in my local config. It is pretty nice how much quieter HTTPS is in the squid logs over HTTP, thanks to it needing to make a direct connect.

**Comment by busterb** [comment]
Thanks again, and sorry for the delay here.

**Comment by OJ** [comment]
Thanks mate!

**Pull request merged**